[Team-11] 새로운 이슈 만들기 기능 추가, DIContainer 에 대한 고찰

Jinsujin opened the pull request:
## 🙋🏻‍♀️ 리뷰어님께 🙋🏻
안녕하세요, 강운! DIContainer 에서 여러가지 문제에 봉착해 고민하고 있습니다. Container 를 활용해서 새로운 기능에 쓰고 싶은데, 이 부분이 어려운것 같습니다. 문제점을 정리해 봤는데, 좋은 아이디어 있으시다면 힌트주시면 감사하겠습니다...🙏🏻

## 📚 작업 사항
- [x] 이슈 만들기 API
- [x] 이슈 만들기 화면에서 API 호출하기
- [x] 이슈를 만들기 위해 사용자의 Repository 선택하는 화면 구현
- [x] 이슈를 만들고 난 후, 목록 화면으로 돌아와 화면 갱신하기

# 🌤 고민과 해결

## 01. Container 에서의 토큰 관리

![Container를사용해ViewController만드는흐름](https://user-images.githubusercontent.com/12508578/176394217-98247cbe-82b0-49f1-8bb5-4b03592eaf62.png)


### `문제 배경`: **Container 의 역할**

- 필요한 객체를 주입하여 화면을 생성해 반환한다.
- AppDelegate 에서 로그인 여부에 따라서 어떤 VC를 rootViewController로 할지 결정한다
- ViewController 에서 화면 이동을 위해, 이동할 화면의 인스턴스 생성을 요청한다.

```swift
// func application(_didFinishLaunchingWithOptions:)
self.window = UIWindow(frame: UIScreen.main.bounds)
window?.rootViewController = container.buildRootViewController()
```

### 하고 싶은 방향

- Container 가 ViewController 를 생성할때, 필요한 객체(이를 테면 Model), 프로퍼티(token) 을 함께 주입하고 싶다.
- 화면을 생성할때 API request 에 필요한 token 값도 Container 에서 관리를 하고 싶다.

### 의문

- 토큰이 필요한 쪽에서(이를테면 API 를 호출하는 객체) 직접 UserDefaults 에 접근해 매번 토큰의 유효성을 검사하는게 좋을 수도 있을 것 같다
- 왜냐하면 Container에서 뷰컨 만드는 시점에 토큰을 넣어주면, 뷰컨 입장에서는 시간이 지나 토큰이 만료되어도 다시 받아올 방법이 없기 때문이다.
- 하지만, 토큰을 가져오는 코드가 여기저기 분산되는 단점이 있어 고민된다. 가장 좋은 방법은 Container 에서만 관리하는 방법일 것이라고 생각되어 진다.

### 🤔지금 문제

- DIContainer에서만 `AccessToken`에 접근하고, VC 생성 시 토큰이 필요한 VC에만 넣어 주고 싶다.
- `UserDefaults.standard.string(forKey: **self**.key)`의 리턴타입이 String?이라서 옵셔널 처리를 해주어야 한다
- 그런데 guard let으로 토큰값의 옵셔널을 벗겨서 주려니, else return 부분에서 뭘 리턴해줘야 하나? 빈 VC를 리턴할 수도 없고..ㅠ

```swift
// DIContainer
// 외부에서 ViewController 인스턴스를 만들기 위해 호출하는 함수.
// 호출하는 곳에서 토큰을 값으로 넣어줘야 한다.
func buildViewController(_ screen: Screen) -> UIViewController {
    switch screen {
// ...
    case .optionSelect(let token, let repositories):
        return OptionSelectViewController(token: token, options: repositories)
    }
}
```

- ViewController 에서 항상 singleton 객체에 직접 접근해 token 값을 얻어와야 하므로, 불필요한 코드가 늘어난다고 생각된다.

```swift
// ViewController
guard let appdelegate = UIApplication.shared.delegate as? AppDelegate,
      let token = GithubUserDefaults.getToken() else {
    return
}

....
guard let viewController = appdelegate.container.buildViewController(														 
  .optionSelect(token: token, repositories: repositoryList)) 
	as? OptionSelectViewController else {
    return
}
```

## 02. 이슈 목록 불러오기

### ❓❓❓현재 문제🤔🤔🤔

이슈 생성은 되었는데, 새로 생성한 이슈가 안보임.

→ 조회가 안되나? 해서 이슈목록 화면 갱신 코드를 점검함 - 정상임.

→ Postman으로 해도 안되나? 했는데 안보임.

→ 우리의 이슈 조회 API가 ****`List issues assigned to the authenticated user` 인데 assigned를 지정하지 않아서 안보이나? 해서 assignees를 현재 로그인한 유저로 넣어줌. 깃헙홈페이지에서 확인했을 때 이슈도 정상 생성되고, 로그인한 유저가 assignee로 할당되어 있음에도 해당 이슈가 보이지 않음.**

→ 유저에 대해 할당된 유저이슈를 조회해야 하는 건가 해서 공식문서의 다른 API인 ****`List user account issues assigned to the authenticated user` 를 사용해 호출해 보았는데도 우리가 만든 이슈는 보이지 않음.**

→ 다만 ****`List repository issues` API를 사용해, 저장소의 owner와 repository를 지정한 상태로 호출했을 때는 우리가 만든 이슈가 정상적으로 노출됨.**

→ 방금 마주한 문제라 해결 방향을 어떻게 잡을지는 아직 상의가 필요합니다😢

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: 새 이슈 생성 시 선택 옵션을 Option enum으로 분리

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: 새 이슈 생성 시 저장소 선택 화면 구현

Jinsujin committed:
[Jinsujin/issue-tracker#20] Fix: issue list api

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: Repository 엔티티 추가 및 repository 요청 코드 수정

Jinsujin committed:
Merge branch 'feat#20' of https://github.com/Jinsujin/issue-tracker into feat#20

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: OptionSelectViewController 생성을 Container에서 담당하도록 변경 (작업중)

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Refactor: Container에서 OptionSelectViewController 생성 (오류 수정)

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: 새 이슈 생성 시 API 요청하여 Repository 불러오기 기능 구현

Jinsujin committed:
[Jinsujin/issue-tracker#20] Feat: 이슈 만들기 버튼과 API 연결

bibi6666667 committed:
[Jinsujin/issue-tracker#20] Feat: 새 이슈 생성 기능 구현

bibi6666667 committed:
[Jinsujin/issue-tracker#20] chore: 불필요한 코드 및 주석 제거

GangWoon commented on IssueTracker/IssueTracker/AppDelegate.swift:
```diff
@@ -11,10 +11,10 @@ import UIKit
-    
     var window: UIWindow?
     
+    let container = Container()
```
프라이빗을 푸신 이유가 싱글톤에서 접근하게 만들기 위해서 일까요??

bibi6666667 replied:
앗..흔적이 남아있네요😅 
Container를 써야 하는 곳은 많은데 객체는 하나만 만들고 싶어서, 싱글톤을 적용해보려 했다가
지난 PR에 static 객체에 접근하기보다는 다른 방법을 사용해보라고 하셨던게 생각나서 싱글톤 없이 해보려고 방향을 바꾸었습니다.
그러다가 (바로 아래와 같은) 더 안 좋은 상황이 발생한 것 같네요..🥲

GangWoon commented on IssueTracker/IssueTracker/Issue/IssueViewController.swift:
```diff
@@ -64,7 +62,14 @@ final class IssueViewController: UIViewController {
     
     @objc func touchedAddButton() {
-        self.navigationController?.pushViewController(Container().buildViewController(.newIssue), animated: true)
+        guard let appdelegate = UIApplication.shared.delegate as? AppDelegate else {
```
이슈 뷰컨트롤러에서 앱에서 단하나만 존재하는 앱델리게이트 객체를 오브젝트화 시키고 있습니다.
윈도우를 넘기셨을때 드렸던 피드백이 기억나실까요??
해당 접근은 제 기준에서, 너무나도 많이 위험해보입니다.

bibi6666667 replied:
음.. 앱델리게이트는 앱이 죽을때까지 살아있는데 그걸 오브젝트로 만들어서 위험하겠군요.ㅠㅠ
Container를 하나만 만들고, 만들어 둔 곳에서 참조해서 쓰려고 이런 식으로 접근했는데 다른 방법을 찾아보겠습니다.

GangWoon commented on IssueTracker/IssueTracker/Issue/IssueViewController.swift:
```diff
@@ -64,7 +62,14 @@ final class IssueViewController: UIViewController {
+            return
+        }
+        self.navigationController?.pushViewController(viewController, animated: true)
+        viewController.delegate = self
```
추가하신 일련의 작업을 글로 작성해주실 수 있을까요??

해당 부분을 작성해주신다면, 제 시야에선 어떤식으로 접근할 것인지 같이 보면 좋을 거 같네요 :)

사실 해당 작업은 제가 좋아하는 선배 개발자분과 제가 같이 했었던 작업인데 여러분들과 할 수 있는 기회가 많아서 개인적으론 매우 기쁩니다.

bibi6666667 replied:
`touchedAddButton()` 말씀하시는 것 같아서 이 부분 글로 적어보겠습니다!
- 이슈목록(IssueViewController)에서 `+`버튼을 누르면 해당 메서드가 실행된다
- 메서드 실행 시 AppDelegate의 Container에 접근해, 새 이슈 화면(NewIssueViewController)를 생성해 가져온 다음,
- 이슈목록(IssueViewController)의 navigationViewController에 새 이슈 화면(NewIssueViewController)를 push해 보여준다.
- 마지막으로 push로 보여지는 새 이슈 화면(NewIssueViewController)의 delegate를 IssueViewController로 설정한다.

GangWoon replied:
해당 부분에서 제가 보이는 시야는
AppDelegate에 포함된 객체를 접근하는게 아닌, 일련의 동작을 주입시켜주며 처리하는 방향이 보입니다.
그렇게 됐을때, IssueViewController는 AppDelegate, 화면 전환 로직을 모두 제거할 수 있게 됩니다.

GangWoon commented on IssueTracker/IssueTracker/Issue/NewIssueViewController.swift:
```diff
@@ -129,18 +161,53 @@ class NewIssueViewController: UIViewController {
-            //2-2. 이슈 목록 조회 다시해서 보여주기
-        //3. api 가 실패했다면 => issue 생성실패 얼럿띄우기
-        
+        guard let token = GithubUserDefaults.getToken(),
+            let selectedRepo = selectedRepo else {
+            return
+        }
+        guard let titleString = self.titleField.text,
+           !titleString.isEmpty else {
+            // TODO: - 타이틀 입력 값이 없다 => 얼럿
+            return
+        }
+
+        service.createIssue(title: titleString, repo: selectedRepo, accessToken: token) { boolResult in
+            if boolResult {
+                self.navigationController?.popViewController(animated: true)
+                self.delegate?.created()
+            }
```
해당 부분 코드도 성격이 유사합니다.
코쿼분들의 코드를 보면, 성격을 나누는거에 포인트를 많이 두세요. 이상하다는건 아닙니다.
다만 이를 사용해서 자연스럽게 붙이는 작업도 나누는거 만큼 중요하다고 생각합니다.

해당 부분도 글로 정리해주시겠나요?

코맨트에 정리하신건 알지만, 다시한번 부탁드리겠습니다. 유사하게 작성하셔도 좋아요 :)

bibi6666667 replied:
'성격을 나눈다'는 게 어떤 의미로 해주신 말씀인가요? `guard let token = ...`, `guard let titleString = ...`과 같은 부분일까요?
일단 `touchedCreateButton()` 이부분도 글로 정리해 보겠습니다!
- 새 이슈 생성 화면에서 '저장'버튼을 클릭하면
- 앱에 저장된 AccessToken과, 현재 선택된 저장소를 가져온다
- 현재 입력된 이슈 제목 값을 가져온다 (제목이 비어있으면 return)
- 제목, AccessToken, 선택된 저장소 정보를 가지고 Service를 통해 API에 이슈 생성 요청을 보낸다.
  - 이 결과가 true이면, navigationStack의 직전 화면으로 돌아가고,
  - 연결된 delegate에 이슈가 생성되었음을 알린다.

GangWoon replied:
객체가 담당하는 역할에 대해서 분리하는 걸 뜻했습니다.
작게 작게 나누는거에 초점을 두다 보니, 해당 자원이 필요할때, 막상 이상하게 접근하는 경우가 많았습니다.

NewIssueViewController의 경우도, 위 행동을 추상화 시킨다면, 버튼이 눌렸을때, 행동을 실행시키며
해당 행동은 NewIssueViewController를 만들 때 넣어주기만 하면 됩니다.

GangWoon commented on IssueTracker/IssueTracker/Issue/OptionSelectViewController.swift:
```diff
@@ -1,11 +1,30 @@
     
+    init(token: String, options: [Repository]) {
+        super.init(nibName: nil, bundle: nil)
+        self.token = token
```
토큰이 정말 필요한 값이였을까요??
사실 이부분도 다른 관점이 시작됩니다.
필요한걸 줄수만 있다면 과연 토큰이 필요할지 고민해보셨으면 좋겠습니다.

bibi6666667 replied:
제 생각에는 OptionSelectViewController에서 필요한건 [Repository] 인데 그럼 이것만 넣어주면 토큰이 필요없어지긴 할 것 같습니다!
그런데 그러면 OptionSelectViewController를 만들어주는 Container에서 API요청까지 해서 [Repository]를 넣어줘야 하는데, 컨테이너가 이 역할까지 해줘도 괜찮은 건가요?🤔

GangWoon replied:
ViewController는 즉 View의 경우 이벤트를 발생시키는 역할만 충실히 만족하며,
해당 부분 외의 작업은 다른 영역에서 담당하는게 유지보수 및 테스트하기 용이합니다.
위 작업을 진행하기 위해서 Container 즉 DIContainer에서 해당 작업을 처리하는게 어색하게 느껴지지 않습니다.
반대로 ViewController를 생성하는데 토큰이 필요한 점이 더 어색하게 느껴집니다.

GangWoon commented on IssueTracker/IssueTracker/Issue/OptionSelectViewController.swift:
```diff
@@ -32,20 +51,31 @@ class OptionSelectViewController: UIViewController {
+        }
+        let selectedItem = options[indexPath.row]
+        delegate?.selected(item: selectedItem)
+        self.navigationController?.popViewController(animated: true)
```
코디네이터 즉 컨테이너를 둔 상태에서 뷰컨끼리의 이동은 전부 추상화 될 수 있는 부분입니다.

GangWoon reviewed (commented):
유사한 패턴으로 문제가 발생하고 있습니다.
왜 불편한지에 대해서 느끼신다는 거 자체가 대단하신거 같네요.

이를 해결하는 방법은 필요한게 무엇인지 분리시키고 넣어줄 수 있는 부분들은 넣어주는 거 입니다.

뜬구름처럼 보이실 수 있지만, 전 다알려드렸습니다 ㅎ ㅡ ㅎ 화이팅이에요 ㅋㅋㅋ

Jinsujin committed:
[Jinsujin/issue-tracker#24] Ref: 옵션 선택하는 화면에서 API 를 호출하도록 수정

- OptionSelectViewController 를 재사용 하기 위함
- 어떤 옵션 타입인지에 따라 다른 데이터를 불러오게 하기 위함

Jinsujin committed:
[Jinsujin/issue-tracker#24] Ref: 선택한 아이템을 이전화면으로 넘길때 리스트에서 option 을 찾아서 변경

Jinsujin committed:
[Jinsujin/issue-tracker#24] Feat: 로그인 후, 첫 화면을 repo list 로 변경

Jinsujin committed:
[Jinsujin/issue-tracker#24] Feat: 이슈 목록을 불러올때, issue tab 에서 생성한 이슈만 가져오도록 함

Jinsujin committed:
Ref: Container 가 token 을 주입받도록 수정

Jinsujin committed:
Ref: Container 에서 ViewController 인스턴스 생성시, token 을 주입하도록 함

Jinsujin committed:
Feat: 레포지토리 리스트 화면에서의 Model 생성과 연결

Jinsujin committed:
Ref: navigation 의 largetitle 설정

bibi6666667 committed:
[Jinsujin/issue-tracker] Feat: NewIssueModel 계층 추가

bibi6666667 committed:
[Jinsujin/issue-tracker] chore: 불필요한 코드 및 주석 삭제

bibi6666667 committed:
[Jinsujin/issue-tracker] OptionSelectModel 계층 추가

bibi6666667 committed:
[Jinsujin/issue-tracker] Feat: 새 이슈 생성 시 Label 붙이는 기능 추가

Jinsujin committed:
Feat: 선택한 옵션이 무엇인지에 따라서 다른데이터를 불러오도록 수정

Jinsujin committed:
Feat: assignees 를 불러오는 API작업중

bibi6666667 committed:
[Jinsujin/issue-tracker] Feat: 새 이슈 생성 시 해당 저장소에 대한 Assignee 조회 구현

bibi6666667 committed:
[Jinsujin/issue-tracker] Feat: 제목, 내용, label, milestone, assignee를 포함해 이슈를 생성하는 기능 구현

bibi6666667 committed:
[Jinsujin/issue-tracker] 불필요한 코드, 주석 삭제 및 파일 분리

Jinsujin commented:
## Coordinator 적용 계획

### 문제 정의

- ViewController 가 DIContainer 에게 화면 이동을 요청하기 위해서, DIContainer 인스턴스를 가지고 있는 AppDelegate 에게 접근할 필요가 있었다.
- AppDelegate 는 앱이 시작하고 끝날때까지 계속해서 살아있는 객체이므로, AppDelegate 에 직접 접근하는 것은 위험할 수 있다.

### 계획

ViewController 가 AppDelegate 에 직접 접근하지 않고, 다음화면을 생성하고 이동할 수 있도록 하기 위해 **Coordinator** 객체를 만든다.

- DIContainer를 사용해 화면이동을 하는 객체-**Coordinator** 를 구현한다
- **Coordinator 는** ViewController 에서 보내는 화면이동 요청을 받아서 화면 이동을 하는 객체이다

![Coordinator 를 사용해 화면이동을하는 흐름](https://user-images.githubusercontent.com/12508578/176815131-b4fc39df-b35a-4bcd-86e3-86d2b66be3ea.png)


**IssueList** 화면에서 **NewIssue** 화면으로 이동하려고 한다. 이때의 화면 이동 흐름은 다음과 같다:

1. **IssueList** 화면에서 delegate 를 통해 `다른화면이동(NewIssue)` 이벤트를 보낸다. 
    - 이때, 이벤트를 보낼 대상은 알지 못한다
2.  **Coordinator** 에서 `화면이동` 이벤트를 받는다. 
3. **Coordinator** 는 **DIContainer** 에게 이동할 화면에 대한 인스턴스 생성을 요청하고 받는다. 
4. **Coordinator** 는 현재 화면을 `NewIssue` 로 변경한다

Jinsujin commented:
DIContainer 를 사용하는 Coordinator 객체를 만들계획을 짜보았습니다. (아직 구현은 못했지만...ㅜ0ㅜ)
issue 생성기능을 마무리했고, 앞으로 남은 작업에 대해 계속 진행해볼 예정입니다
### 계속해볼 작업들:
- [ ] Unit Test
- [ ] 클로저의 self 사용과 순환참조 알아보기 (메모리 누수)
- [ ] 접근성 구현하기
- [ ]  Alamofire, Snapkit 프레임워크 파헤쳐보기

GangWoon commented on IssueTracker/IssueTracker/Container.swift:
```diff
@@ -29,10 +30,14 @@ struct Container {
+            return viewController
+        case .repos(token: let token):
+            let viewController = ReposViewController(token: token)
+            viewController.title = "Repos"
```
ViewController에 대한 뷰를 꾸미는 작업 내부에서 처리하는게 더 좋아보입니다.
외부에서 ViewController 를 꾸밀때는, 더 의미가 명확한 작업을 하셨으면 좋겠습니다.

GangWoon commented on IssueTracker/IssueTracker/Repos/ReposViewController.swift:
```diff
@@ -0,0 +1,96 @@
+    
+    private let service = IssueService()
+    private let token: String
+    private let tableViewCellIdentifier = "tableViewCellIdentifier"
```
유의미한 코드일까요?

GangWoon commented on IssueTracker/IssueTracker/Repos/ReposViewController.swift:
```diff
@@ -0,0 +1,96 @@
+            case .failure(let error):
+                print(error)
+            }
+        }
```
UserDefault 객체 하나와 Service 객체 하나 총 두가지 구체 타입을 ViewController에게 노출된 상황입니다.
해당 상황에서 행위를 추상화 시킬 수 있다면 ViewController가 해당 두 구체 타입에 대해서 인지할 이유가 있을까요?

위 작업을 진행했을때, 어떤 점이 이득으로 다가올까요?

GangWoon commented on IssueTracker/IssueTracker/AppDelegate.swift:
```diff
@@ -13,7 +13,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     var window: UIWindow?
     
-    let container = Container()
+    let container = Container(token: GithubUserDefaults.getToken())
```
Container 객체도 싱글톤으로 관리하고 있지 않는 상태입니다.
GithubUserDefaults도 동일하게 관리할 수 없을까요?

Jinsujin replied:
리펙토링 하다보니 GithubUserDefaults 객체를 사용하는 곳이 AppDelegate 밖에 없네요! static 을 제거하도록 하겠습니다

GangWoon commented on IssueTracker/IssueTracker/Repos/ReposViewController.swift:
```diff
@@ -3,18 +3,17 @@ import SnapKit
     
-    init(token: String) {
-        self.token = token
+    init(service: IssueService) {
```
코드의 위치를 옮기시면서, 토큰 값을 숨기는 효과는 저희가 누린거 같습니다.
다만 ViewController에서 발생하는 이벤트를 핸들링한다는 관점에서 접근할때, 결과적으로 토큰에 대한 디펜던시가 사라졌을까요?

Jinsujin replied:
`ViewController -> Model -> Service(init: token)`
- ViewController 가 Model 에 의존하고 있고, Model 이 token 을 사용하는 Service 에 의존하고 있으니 결과적으로 `ViewController  token 을 의존하고 있다` 라는 것일까요?
- API 요청을 하기위해 Service 객체는 token 을 알아야 하는데, 그럴러면 token 은 필수값이라 생각이 되어집니다
- token 에 대한 의존성을 줄일 수 있다는 말씀 이실까요?

GangWoon commented on IssueTracker/IssueTracker/Repos/ReposViewController.swift:
```diff
@@ -3,34 +3,28 @@ import SnapKit
-                self?.options = repositoryList
+        
+        model.fetchViewData()
+        model.updated = { [weak self] repos in
```
외부에서 모델과 뷰컨을 모두 생성하는 시점이 존재합니다.
이를 뷰컨에서 정립하는게 아닌, 해당 시점에서 처리하는 건 어떨까요?

Jinsujin replied:
그렇게 하는게 좋겠네요! 수정 하도록 하겠습니다

GangWoon commented on IssueTracker/IssueTracker/OptionSelect/OptionSelectViewController.swift:
```diff
@@ -53,6 +40,12 @@ class OptionSelectViewController: UIViewController {
     
+    private func bind() {
+        model.updatedOptions = {
+            self.tableView.reloadData()
```
클로저 캡처는 매순간 조심하셔야 합니다.

Jinsujin committed:
Ref: GithubUserDefaults 의 static 제거

Jinsujin committed:
Ref: Repos 바인딩을 Container 에 하도록 변경

Jinsujin committed:
Doc: readme.md 업뎃

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Container의 의존성을 관리하는 ContainerEnvironment 추가

bibi6666667 committed:
[Jinsujin/issue-tracker #28] ReposModel의 의존성인 IssueService.requestRepos()를 위해 service 전체가 아닌 클로저만 Environment로 받도록 변경

bibi6666667 committed:
[Jinsujin/issue-tracker #28] ReposModel의 의존성인 IssueService.requestRepos()를 위해 service 전체가 아닌 클로저만 Environment로 받도록 변경

bibi6666667 committed:
Revert "[Jinsujin/issue-tracker #28] ReposModel의 의존성인 IssueService.requestRepos()를 위해 service 전체가 아닌 클로저만 Environment로 받도록 변경"

This reverts commit 2cf9ed00cd719fa1a533869dc18c8e077bdf6262.

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 IssueModel이 service 전체를 받지 않고, service에서 사용하는 클로저만 environment로 받도록 변경

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Feat: NewIssueModel에 service 전체 대신, 필요한 클로저만 environment로 주입

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Feat: OptionSelect에서 사용하는 service의 메서드들을 클로저로 주입하도록 변경

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Refactor: LoginVC와 LoginModel 분리 및 OAuthService에서 필요한 클로저만 갖도록 환경 분리

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Refactor: IssueVC 생성 시 필요한 작업을 viewDidLoad대신 Container에서 하도록 변경

bibi6666667 committed:
[Jinsujin/issue-tracker] #28 Refactor: OptionSelectVC 생성 시 필요한 설정 부분을 viewDidLoad 대신 Container에서 처리하도록 변경

bibi6666667 committed:
Merge branch 'develop' into bibi/refactor

bibi6666667 committed:
Merge pull request #29 from Jinsujin/bibi/refactor

bibi의 리팩토링 PR😎

bibi6666667 committed:
Feat : AppDelegate를 위한 코디네이터인 AppCoordinator 생성 및 실행 시점 rootVC 판단 로직 분리

bibi6666667 committed:
Feat: Container에 register, resolve 패턴 적용해 Login/Repos 화면 전환

GangWoon commented on IssueTracker/IssueTracker/Repos/Model/ReposModel.swift:
```diff
@@ -2,38 +2,52 @@ import Foundation
     var updated: (([Repository]) -> Void)?
     
-    private var optionViewData: [Repository] {
+    private var ReposList: [Repository] {
```
카멜케이스는 유지 하셨으면 좋겠습니다.

GangWoon commented on IssueTracker/IssueTracker/NewIssue/Model/NewIssueModel.swift:
```diff
@@ -9,15 +9,19 @@ import Foundation
 }
+
+struct NewIssueModelEnvironment {
+    let createIssue: (String, Repository, String, Label?, Milestone?, Assignee?, @escaping (Bool) -> Void) -> Void
```
타입이 길어지기 시작하면 typealias를 사용하셔서 의미를 부여하면 좋을 거 같습니다.

GangWoon replied:
아니면 의미가 있는 모델로 묶는 방법도 있습니다.

GangWoon commented on IssueTracker/IssueTracker/Container.swift:
```diff
@@ -35,18 +35,24 @@ class Container {
             let viewController = IssueViewController(model: model, repo: selectedRepo) // Issue -> Issues
             return viewController
         case .newIssue(let repo):
             let model = NewIssueModel(environment: .init(createIssue: environment.issueService.createIssue(title:repo:content:label:milestone:assignee:completion:)))
```
클로저 주입할때, 놓치기 쉬운 부분입니다.
해당 Environment를 생서할 때, 보이지 않게 self.environment를 주입한 시점입니다.
즉 self를 강하게 캡쳐시켰기 때문에 리테인 사이클이 발생합니다.

bibi6666667 committed:
Feat: ReposVC에 코디네이터와 컨테이너 적용

bibi6666667 committed:
Feat: ReposViewController를 보여줄 때 ReposCoordinator를 사용하도록 변경

bibi6666667 committed:
Feat: IssueViewController 화면 흐름을 관리하는 IssueCoordinator 추가 #28

bibi6666667 committed:
Feat: NewIssueViewController를 관리하는 NewIssueCoordinator 추가 및 새 이슈 생성에 필요한 NewIssueFormat 구조체 추가

bibi6666667 committed:
Feat: 모든 뷰컨에 coordinator 프로퍼티 추가 및 NewIssue 생성 시 뒤로가기 동작 구현 #28

bibi6666667 committed:
Feat: OptionSelectViewController를 관리하는 OptionSelectCoordinator 추가 #28

bibi6666667 committed:
Feat : coordinator생성 시 register한 뒤, childCoordinators 관리할 때 활용 #28

기존 VC에서 coordinator를 직접 소유하던 부분을 삭제하고
AppCoordinator에서 childCoordinators를 관리할 때 register된 coordinator 사용

bibi6666667 committed:
Feat : 이슈생성, 옵션선택 후 coordinator에서 이전 화면으로 되돌아가며 reload호출 (동작하지 않음) #28

bibi6666667 committed:
Feat : OptionSelectVC를 coordinator로 연결 - 옵션마다 다른 화면 생성 #28

bibi6666667 committed:
Merge branch 'bibi/refactor' of https://github.com/Jinsujin/issue-tracker into bibi/refactor

# Conflicts:
#	IssueTracker/IssueTracker/AppDelegate.swift
#	IssueTracker/IssueTracker/Issue/IssueViewController.swift
#	IssueTracker/IssueTracker/NewIssue/NewIssueViewController.swift
#	IssueTracker/IssueTracker/Repos/ReposViewController.swift

bibi6666667 committed:
Feat : 이슈 생성 요청과 이슈목록 요청 사이에 딜레이 추가, 받아온 이슈에 새로 생성한 이슈가 없다면 재요청 #28

bibi6666667 committed:
Chore : 미사용 메서드와 주석 삭제 #28

bibi6666667 committed:
Chore : 불필요한 주석 및 코드 삭제 #28

bibi6666667 committed:
refactor: ViewController deinit이 호출되지 않는 문제 디버깅 #28
- delegate 프로토콜을 weak var로 갖도록 변경
- 클로저 내에 선언된 self 를 weak self로 변경

bibi6666667 committed:
refactor: 변수명을 적절하게 수정 #28

bibi6666667 committed:
refactor: 불필요한 self 키워드 삭제 #28

bibi6666667 committed:
refactor: 이슈 생성 요청 이후 새 이슈가 생성되었는지 확인한 다음 이슈 목록으로 되돌아가도록 변경 (뷰 미완성) #28

bibi6666667 committed:
refactor : 새 이슈 생성 -> 이슈 조회 흐름 수정. #28
- 새 이슈 생성 시, "이슈생성 요청" 이후 "이슈목록 조회 요청" 을 보내 생성된 이슈가 반영될때까지 시간 간격을 두고 재요청을 한다. 이 때 indicator를 나타낸다.
- 반영이 되면 이전 화면인 이슈목록으로 돌아가 이슈목록을 조회하고, 화면을 갱신한다.

bibi6666667 committed:
debug : 로그인 직후 앱으로 돌아왔을 때 저장소 목록이 나오지 않는 현상 디버깅(해결) #28

bibi6666667 committed:
refactor : Container에서 뷰컨트롤러와 관련 코디네이터를 묶어 관리하고 삭제 작업도 처리하도록 변경 #28

bibi6666667 committed:
refactor : 뷰컨트롤러와 관련 코디네이터 이름을 담은 ViewControllerCoordinator 열거형을 사용해 중복 코드 개선 #28

bibi6666667 committed:
chore : 불필요한 주석 및 print문 삭제 #28

bibi6666667 committed:
debug : IssueService 초기화를 위해 토큰 기본값을 넘기는 대신 옵셔널로 타입 수정 #28

bibi6666667 committed:
Merge pull request #31 from Jinsujin/bibi/refactor

[Bibi/refactor] DIContainer 및 Coordinator 적용 및 이슈목록 갱신 문제 디버깅
